turbopack/crates/turbo-persistence/src/compaction/selector.rs

```diff
@@ -46,183 +35,327 @@ fn extend_range(a: &mut Range, b: &Range) -> bool {
+        .ranges()
+        .map(|(range, (dup_info, count))| {
+            let duplicated_size = dup_info.duplication(&range);
+            let total_size = dup_info.size(&range);
+            let overlap = spread(&range) as f32 * count.saturating_sub(1) as f32;
+            (duplicated_size, total_size, overlap)
+        })
+        .reduce(|(dup1, total1, overlap1), (dup2, total2, overlap2)| {
+            (dup1 + dup2, total1 + total2, overlap1 + overlap2)
+        })
+        .map(|(duplicated_size, total_size, overlap)| {
+            (
+                duplicated_size,
+                if total_size > 0 {
+                    duplicated_size as f32 / total_size as f32
+                } else {
+                    0.0
+                },
+                overlap,
+            )
+        })
+        .unwrap_or((0, 0.0, 0.0));
+
+    CompactableMetrics {
+        coverage: coverage / full_spread,
+        overlap: overlap / full_spread,
+        duplicated_size,
+        duplication,
     }
-    coverage / spread(&full_range) as f32
 }
 
 /// Configuration for the compaction algorithm.
 pub struct CompactConfig {
     /// The minimum number of files to merge at once.
     pub min_merge: usize,
 
+    /// The optimal number of files to merge at once.
+    pub optimal_merge_count: usize,
+
     /// The maximum number of files to merge at once.
     pub max_merge: usize,
 
     /// The maximum size of all files to merge at once.
     pub max_merge_size: u64,
+
+    /// The amount of duplication that need to be in a merge job to be considered for merging.
+    pub min_merge_duplication_size: u64,
+
+    /// The optimal duplication size for merging.
+    pub optimal_merge_duplication_size: u64,
+
+    /// The maximum number of merge segments to determine.
+    pub max_merge_segments: usize,
 }
 
-/// For a list of compactables, computes merge and move jobs that are expected to perform best.
-pub fn get_compaction_jobs<T: Compactable>(
-    compactables: &[T],
-    config: &CompactConfig,
-) -> CompactionJobs {
-    let (jobs, _) = get_compaction_jobs_internal(compactables, config, 0);
-    jobs
+impl Default for CompactConfig {
+    fn default() -> Self {
+        const MB: u64 = 1024 * 1024;
+        Self {
+            min_merge: 2,
+            optimal_merge_count: 8,
+            max_merge: 32,
+            max_merge_size: 500 * MB,
+            min_merge_duplication_size: MB,
+            optimal_merge_duplication_size: 10 * MB,
+            max_merge_segments: 8,
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+struct DuplicationInfo {
+    total_size: u64,
+    max_size: u64,
 }
 
-fn get_compaction_jobs_internal<T: Compactable>(
+impl DuplicationInfo {
+    fn duplication(&self, range: &Range) -> u64 {
+        if self.total_size == 0 {
+            return 0;
+        }
+        // This is equal to (self.total_size - self.max_size) * spread(range) / u64::MAX
+        // but we use widening_mul to avoid overflow.
```

Comment:
Can we just use `u128` to avoid the nightly API?

It's really confusing from reading this why we need to divide by `u64::MAX`.